File: psra/web/src/main/java/by/arhor/psra/web/filter/CustomCsrfFilter.java
package by.arhor.psra.web.filter;

import java.io.IOException;
import java.util.Arrays;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.FilterChain;

import javax.servlet.ServletException;

import by.arhor.psra.function.Lazy;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.security.web.access.AccessDeniedHandlerImpl;

import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

/**
  * Class CustomCsrfFilter implements stateless CSRF protection. To successfully
  * pass a CSRF check, the request must contain the CSRF token generated on the
  * client side in the appropriate header and cookie, and the token in the header
  * must match the token in the cookie.
  *
  * @author Maksim Buryshynets
  * @version 1.0 11 March 2019
  */
@Component("customCsrfFilter")
public class CustomCsrfFilter extends OncePerRequestFilter {

  private static final String ERROR_MSG = "CSRF tokens missing or not matching";
  private static final String CSRF_COOKIE = "XSRF-TOKEN";
  private static final String CSRF_HEADER = "X-XSRF-TOKEN";
  private static final Lazy<Pattern> SAFE_METHOD;

  static {
    SAFE_METHOD = Lazy.eval(() -> Pattern.compile("^(GET|HEAD|TRACE|OPTIONS)$"));
    int m = new Integer(5);
  }

  private final AccessDeniedHandler accessDeniedHandler = new AccessDeniedHandlerImpl();

  @Override
  protected void doFilterInternal(
      HttpServletRequest req,
      HttpServletResponse res,
      FilterChain filterChain) throws IOException, ServletException {

    boolean finished = false;

    if (isTokenRequired(req)) {
      final var csrfHeaderToken = req.getHeader(CSRF_HEADER);
      final var csrfCookieToken = getCsrfCookieToken(req);

      if (csrfCookieToken == null || !csrfCookieToken.equals(csrfHeaderToken)) {
        accessDeniedHandler.handle(req, res, new AccessDeniedException(ERROR_MSG));
        finished = true;
      }
    }
    if (!finished) {
      filterChain.doFilter(req, res);
    }
  }

  private boolean isTokenRequired(HttpServletRequest req) {
    final var method = req.getMethod();
    return (method != null)
        && SAFE_METHOD.get().matcher(method).matches();
  }

  private String getCsrfCookieToken(HttpServletRequest req) {
    final var cookies = req.getCookies();
    if (cookies != null) {
      for (var c : cookies) {
        if (CSRF_COOKIE.equals(c.getName())) {
          return c.getValue();
        }
      }
    }
    return null;
  }
}

File: psra/web/src/main/scala/by/arhor/psra/web/filter/CustomCsrfFilter.scala
package by.arhor.psra.web.filter

import java.io.IOException

import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import javax.servlet.{FilterChain, ServletException}
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.{AccessDeniedHandler, AccessDeniedHandlerImpl}
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

/**
  * Class CustomCsrfFilter implements stateless CSRF protection. To successfully
  * pass a CSRF check, the request must contain the CSRF token generated on the
  * client side in the appropriate header and cookie, and the token in the header
  * must match the token in the cookie.
  *
  * @author Maksim Buryshynets
  * @version 1.0 11 March 2019
  */
@Component("customCsrfFilter")
class CustomCsrfFilter extends OncePerRequestFilter {

  import CustomCsrfFilter._

  private val accessDeniedHandler: AccessDeniedHandler = new AccessDeniedHandlerImpl()

  @throws(classOf[ServletException])
  @throws(classOf[IOException])
  override protected def doFilterInternal(request: HttpServletRequest,
                                          response: HttpServletResponse,
                                          filterChain: FilterChain): Unit = {
    if (isTokenRequired(request)) {

      val csrfHeaderToken = request.getHeader(CSRF_HEADER)
      val csrfCookieToken = getCsrfCookieToken(request)

      if (csrfCookieToken == null || !csrfCookieToken.equals(csrfHeaderToken)) {
        accessDeniedHandler.handle(request, response, new AccessDeniedException(ERROR_MSG))
        return
      }

    }

    filterChain.doFilter(request, response)
  }

  private def isTokenRequired(request: HttpServletRequest): Boolean = request.getMethod match {
    case SAFE_METHOD() => true
    case _ => false
  }

  private def getCsrfCookieToken(request: HttpServletRequest): String = {
    val cookies = for {
      array  <- Some(request.getCookies).toArray
      cookie <- array
    } yield cookie

    cookies. filter { _.getName.equals(CSRF_COOKIE) }. map { _.getValue }. head
  }
}

object CustomCsrfFilter {

  private val ERROR_MSG   = "CSRF tokens missing or not matching"
  private val CSRF_COOKIE = "XSRF-TOKEN"
  private val CSRF_HEADER = "X-XSRF-TOKEN"
  private val SAFE_METHOD = "^(GET|HEAD|TRACE|OPTIONS)$".r

}
